src/Interpreters/JoinInfo.cpp

```diff
@@ -249,4 +287,201 @@ DataTypePtr JoinActionRef::getType() const
+            "Cannot serialize JoinActionRef({}) because actions_dag is not in the DAGs list. DAG: {}",
+            column_name, actions_dag->dumpDAG());
+
+    writeVarUInt(pos, out);
+    writeStringBinary(column_name, out);
+}
+
+JoinActionRef JoinActionRef::deserialize(ReadBuffer & in, const ActionsDAGRawPtrs & dags)
+{
+    UInt64 pos;
+    readVarUInt(pos, in);
+
+    if (pos >= dags.size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Cannot deserialize JoinActionRef because actions_dag position {} outside of range (0, {})",
+            pos, dags.size());
+
+    JoinActionRef res(nullptr);
+    readStringBinary(res.column_name, in);
+    res.actions_dag = dags[pos];
+
+    return res;
+}
+
+void JoinPredicate::serialize(WriteBuffer & out, const JoinActionRef::ActionsDAGRawPtrs & dags) const
+{
+    serializePredicateOperator(op, out);
+    left_node.serialize(out, dags);
+    right_node.serialize(out, dags);
+}
+
+JoinPredicate JoinPredicate::deserialize(ReadBuffer & in, const JoinActionRef::ActionsDAGRawPtrs & dags)
+{
+    auto op = deserializePredicateOperator(in);
+    auto left_node = JoinActionRef::deserialize(in, dags);
+    auto right_node = JoinActionRef::deserialize(in, dags);
+    return {std::move(left_node), std::move(right_node), op};
+}
+
+static void serializePredicates(WriteBuffer & out, const std::vector<JoinPredicate> & predicates, const JoinActionRef::ActionsDAGRawPtrs & dags)
+{
+    writeVarUInt(predicates.size(), out);
+    for (const auto & predicate : predicates)
+        predicate.serialize(out, dags);
+}
+
+static std::vector<JoinPredicate> deserializePredicates(ReadBuffer & in, const JoinActionRef::ActionsDAGRawPtrs & dags)
+{
+    UInt64 size;
+    readVarUInt(size, in);
+
+    std::vector<JoinPredicate> predicates;
+    predicates.reserve(size);
+
+    for (size_t i = 0; i < size; ++i)
+        predicates.emplace_back(JoinPredicate::deserialize(in, dags));
+
+    return predicates;
+}
+
+static void serializeJoinActions(WriteBuffer & out, const std::vector<JoinActionRef> & actions, const JoinActionRef::ActionsDAGRawPtrs & dags)
+{
+    writeVarUInt(actions.size(), out);
+    for (const auto & action : actions)
+        action.serialize(out, dags);
+}
+
+static std::vector<JoinActionRef> deserializeJoinActions(ReadBuffer & in, const JoinActionRef::ActionsDAGRawPtrs & dags)
+{
+    UInt64 size;
+    readVarUInt(size, in);
+
+    std::vector<JoinActionRef> actions;
+    actions.reserve(size);
+
+    for (size_t i = 0; i < size; ++i)
+        actions.emplace_back(JoinActionRef::deserialize(in, dags));
+
+    return actions;
+}
+
+void JoinCondition::serialize(WriteBuffer & out, const JoinActionRef::ActionsDAGRawPtrs & dags) const
+{
+    serializePredicates(out, predicates, dags);
+    serializeJoinActions(out, left_filter_conditions, dags);
+    serializeJoinActions(out, right_filter_conditions, dags);
+    serializeJoinActions(out, residual_conditions, dags);
```

Comment:
I’m also not sure how this might change with an extended step for multiple inputs, but it should be fine since the predicate will still refer to the two particular tables